### ISSUE TITLE
Meet AA contrast requirements

### ISSUE THREAD

**Issue by gakimball** [zurb/foundation-sites#7387]
I'm combining all of our open issues around contrast into one. Our goal is AA, not AAA.

**Buttons:**
- [ ] [Primary](http://webaim.org/resources/contrastchecker/?fcolor=ffffff&bcolor=2199e8) #7103
- [ ] [Secondary](http://webaim.org/resources/contrastchecker/?fcolor=ffffff&bcolor=777777) #7104
- [ ] [Success](http://webaim.org/resources/contrastchecker/?fcolor=ffffff&bcolor=3adb76) #7376
- [ ] [Alert](http://webaim.org/resources/contrastchecker/?fcolor=ffffff&bcolor=ec5840) #7377
- [ ] [Warning](http://webaim.org/resources/contrastchecker/?fcolor=ffffff&bcolor=ffae00) #7378
- [x] Disabled #7379

Disabled is an open question, as according to the W3C's guidelines, contrast requirements are not necessary for disabled elements.

**If any other components don't meet AA requirements, let's add them here also.**

**Comment by Owlbertz** [comment]
@gakimball the difference between `disabled` and `aria-disabled` should be, that not all elements support `disabled` according to HTML standard and that `disabled` actually disables the element, while `aria-disabled` just indicates a disabled UI component to AT. This means, `buttons` would go with the `disabled` attribute, and `a` would need to use `aria-disabled` or `.disabled`. I feel like we should include all three ways in the CSS selector, and then recommend `disabled` for `button` and `input` in the docs.

**Comment by Owlbertz** [comment]
Another place with too low ratio is the title of Off-canvas list titles: http://webaim.org/resources/contrastchecker/?fcolor=8a8a8a&bcolor=2c2c2c

**Comment by andycochran** [comment]
Also of too low ratio: 
* regular `$primary-color` links
* links in callouts
* `header small`
* `.tabs-title` (especially `.is-active`)
* blockquotes
* `.subheader`

I'm not sure about the requirements for progress bars. It's not text on a background, but the ratio is too low.

**Comment by davefuentes** [comment]
@Owlbertz is right. Disabled elements don't need to meet WCAG criteria. We can skip that one.

**Comment by davefuentes** [comment]
I think if we check colors with backgrounds and foregrounds, we can fix all possible issues:

```
$primary-color: #2199e8;
$secondary-color: #777;
$success-color: #3adb76;
$warning-color: #ffae00;
$alert-color: #ec5840;
$light-gray: #e6e6e6;
$medium-gray: #cacaca;
$dark-gray: #8a8a8a;
$black: #0a0a0a;
$white: #fefefe;
$body-background: $white;
$body-font-color: $black;
```

**Comment by zurbrandon** [comment]
I've taken a stab here:
https://github.com/zurb/foundation-sites/pull/7603

If anyone wants to make a pass and see if they cvan bring some of the vibrance back it would be appreciated. @davefuentes @andycochran @Owlbertz @gakimball

**Comment by davefuentes** [comment]
Can you provide a screenshot between old colors and the new ones you proposed?

**Comment by andycochran** [comment]
Trying to tackle this one. It's tricky. 

Here's the slightest change to the existing palette that would meet AA contrast requirements… __*if*__ the warning button used `$button-color-alt` for its text: 

```sass
$foundation-palette: (
  primary: #1a7cbd,
  secondary: #767676,
  success: #238748,
  warning: #ffae00,
  alert: #cc4b37,
);
```
![image](https://cloud.githubusercontent.com/assets/409279/19833373/4a4f9768-9e0d-11e6-88a0-4732151f078c.png)

The problem is, `@function foreground()` uses `lightness()` to determine if the light or dark text color should be used. [This article](https://madebymike.com.au/writing/accessible-contrast-with-less-and-sass/), which explains how `lightness()` fails at calculating contrast, recommends using a luminance based method — [something like this](https://gist.github.com/voxpelli/6304812) — to calculate contrast ratios. 

It's also worth noting that `$button-color` and `$button-color-alt` must be set to light and dark values respectively for `foreground()` to work. The description of `$button-color-alt` even says, "_Font color for buttons, if the background is light_," which makes the assumption that buttons are by default dark. Perhaps `$button-color-light` and `$button-color-dark` would be more semantic? 

Any thoughts before I reeeeally complicate the `foreground()` function? 

/cc @brettsmason @JeremyEnglert @ncoden @rafibomb

**Comment by ncoden** [comment]
@andycochran hum... Nothing to say, we should use the function of the article instead of `lightness ()`.
`$button-color-{light, dark}` would refer to `light/dark color` or `color for light/dark button` ?

**Comment by andycochran** [comment]
Good point, @ncoden; `$button-color-light` could be misunderstood. Might the default vars be enough context?

```sass
$button-color-light: $white;
$button-color-dark: $black;
```

**Comment by andycochran** [comment]
^ That PR probably isn't quite ready. I'd appreciate any feedback.

**Comment by andycochran** [comment]
@ncoden So, I didn't even use two variables for button colors. We only define one explicit button background color. So we only need to define one button font color. That button's font color can either be set to a specific color or to this function `pick-best-color($button-background)`, which defaults to choosing from `$black` or `$white`: 

```sass
pick-best-color($base: $background, $colors: ($white, $black))
```

…but it can also accept custom vars: 
```sass
pick-best-color($tab-background, ($primary-color, $white))
```

As I'm explaining this, [rubber duck](https://en.wikipedia.org/wiki/Rubber_duck_debugging) is telling me I should amend my PR by keeping `$button-color` and `$button-color-alt` in settings and have `pick-best-color()` default to those instead of putting the function in settings. That way, you'll be able to choose font colors for the buttons generated by `$foundation-palette()`.

**Comment by andycochran** [comment]
Okay, lots of updates on the #9319 PR. Check 'em out.

**Comment by andycochran** [comment]
Closing this as #9319 will give v6.3 accessible colors! Woot.

**Comment by theschmocker** [comment]
> Also of too low ratio:
> 
>     * `header small`
> 
>     * `.tabs-title` **(especially `.is-active`)**
> 
>     * blockquotes
> 
>     * `.subheader`
> 

The default colors for these still fall below WCAG AA requirements. Except for .tab-title.is-active, the culprit is the default $dark-gray. It is too light on the default background color ($white: \#fefefe)

![WCAG Contrast checker - #8a8a8a on #fefefe doesn't meet AA requirement )](https://user-images.githubusercontent.com/26126157/55177287-e011b080-5150-11e9-9d08-2e2793049e12.png)

**Comment by JeremyEnglert** [comment]
@theschmocker it's probably worth opening a separate issue for this as this particular ticket is pretty old.

**Comment by andycochran** [comment]
Drives me nuts that header>small has an explicit color. All my apps override this.
